### ISSUE TITLE
NoMethodError - undefined method `<=' for nil:NilClass

### ISSUE THREAD

**Issue by xmik** [bundler/gemstash#106]
I've been using Gemstash for several months, version 1.0.0. Today I redeployed Gemstash Docker container with version 1.0.2. After some time (1.5 h), I've got an error on `bundle install`: 
```bash
Gem::RemoteFetcher::FetchError: bad response Internal Server Error 500 (http://my_rubygems_cache_public_host/gems/fog-local-0.3.0.gem)
An error occurred while installing fog-local (0.3.0), and Bundler cannot
continue.
Make sure that `gem install fog-local -v '0.3.0'` succeeds before bundling.
```

The logs say (and they are upside down):
```
[2016-09-24 19:45:24 +0000] - INFO - [7] 10.1.1.224 - - [24/Sep/2016:19:45:24 +0000] "GET /gems/fog-local-0.3.0.gem HTTP/1.1" 200 - 1.4742
/usr/local/bundle/gems/puma-2.16.0/lib/puma/thread_pool.rb:106:in `block in spawn_thread'
/usr/local/bundle/gems/puma-2.16.0/lib/puma/server.rb:270:in `block in run'
# ... shorthened
/usr/local/bundle/gems/gemstash-1.0.2/lib/gemstash/storage.rb:274:in `load_properties'
/usr/local/bundle/gems/gemstash-1.0.2/lib/gemstash/storage.rb:279:in `check_resource_version'
Did you mean? <=>:
[2016-09-24 19:45:24 +0000] - ERROR - 2016-09-24 19:45:24 - NoMethodError - undefined method `<=' for nil:NilClass
[2016-09-24 19:45:24 +0000] - INFO - [7] 172.17.0.1 - - [24/Sep/2016:19:45:24 +0000] "GET / HTTP/1.1" 302 - 0.0008
[2016-09-24 19:45:23 +0000] - INFO - Gem fog-local-0.3.0 is not cached, fetching gem
[2016-09-24 19:45:23 +0000] - INFO - [7] 10.1.1.224 - - [24/Sep/2016:19:45:23 +0000] "GET /gems/fog-google-0.1.0.gem HTTP/1.1" 200 - 1.7989
[2016-09-24 19:45:22 +0000] - INFO - [7] 10.1.1.224 - - [24/Sep/2016:19:45:22 +0000] "GET /gems/fog-google-0.1.0.gem HTTP/1.1" 200 - 1.6589
[2016-09-24 19:45:22 +0000] - INFO - [7] 10.1.1.224 - - [24/Sep/2016:19:45:22 +0000] "GET /gems/fog-google-0.1.0.gem HTTP/1.1" 200 - 1.6708
[2016-09-24 19:45:21 +0000] - INFO - Gem fog-google-0.1.0 is not cached, fetching gem
```

Don't you think there should be some check that version is not nil? https://github.com/bundler/gemstash/blob/master/lib/gemstash/storage.rb#L279

**Comment by smellsblue** [comment]
@xmik Sorry for the troubles in upgrading!

Have you been using the gem version of gemstash or a checkout of the repository?

It *should* be impossible for the version to be nil, as we have been storing the version information since `1.0.0`: https://github.com/bundler/gemstash/blob/v1.0.0/lib/gemstash/storage.rb#L307

However, if you have been using a checkout of the repository, it's possible you were using a version before I added the versioning code.

Were you using version `1.0.0` or one of the pre-release versions? Looking through the tags, it looks like that has been around since `1.0.0.pre.2`, so if you were using `1.0.0.pre.1`, then that would explain the issue.

Regardless, I think you are right... some kind of fallback for a corrupted properties file would be nice... but I'm not entirely sure what the best answer would be.... stash the gem in a special backup location and then throw it away and re-fetch it...? Silently ignoring and just trying to use the gem as is would work fine now since we haven't made any real changes to the storage engine, but it could be problematic in future versions if we make changes.

**Comment by smellsblue** [comment]
@xmik I think this would only happen for private gems for `1.0.0.pre.1`, and I had written a script you can run to fix gems that you can see from the [changelog](https://github.com/bundler/gemstash/blob/master/CHANGELOG.md#upgrade-notes-1). You can see the script here: https://gist.github.com/smellsblue/53f5a6757dcc91ad10bc

**Comment by xmik** [comment]
@smellsblue, thanks for replying.

Earlier I was using a docker image: https://hub.docker.com/r/artemave/gemstash/ , in [Dockerfile](https://github.com/artemave/gemstash-docker/blob/master/Dockerfile) it installed Gemstash with:
```
gem install --no-ri --no-rdoc gemstash
```
In order to be able to rely that the docker image had always the same Gemstash version: `1.0.0`, I pulled it into a private docker registry. 

Then, after I upgraded bundler (in many our projects) from `1.10.5` to `1.12.5`, Gemstash started working really slow, but it was fixed by https://github.com/bundler/gemstash/issues/100 in Gemstash `1.0.2`. So now I am using new docker image (I made a PR to artemave/gemstash: https://github.com/artemave/gemstash-docker/pull/1) and I really have to use Gemstash `>= 1.0.2`.

### Potential solution
> stash the gem in a special backup location and then throw it away and re-fetch it

I'd say that it would be even ok to re-fetch it and, if it succeeded, remove the old cached version of the gem. (And this does not concern private gems, because if they were stored only in Gemstash, then there is no place to re-fetch them from). Btw, I have never used Gemstash for private gems.

However, for the last ~11 days I've been using Gemstash heavily and successfully (on workstation and on ci agents) and this error did not occur any more (for any other or the same fog-local gem). Between redeployment (from `1.0.0` to `1.0.2`) I didn't clean up Gemstash storage directory and I think this is the reason of my error. So feel free to close this issue without any changes in code.

**Comment by smellsblue** [comment]
@xmik hmmm, I'm wondering if we have a race condition.... is it possible you were fetching the same gem for the first time from 2 different servers at about the same time? That might explain a random one-off error like that....

**Comment by xmik** [comment]
I can't find that ci job output anymore, but it is definitely possible that 2 different servers were fetching the same gem (with the same version) using Gemstash cache at the same time.

**Comment by pcarranza** [comment]
The only way I got an error like this was when I had an old version of gemstash where I had an old version of the properties file, for this I created this PR => https://github.com/bundler/gemstash/pull/87 which is not merged yet.

I think that this is a transient error and there is little else to do here than what is proposed in that error message.

**Comment by smellsblue** [comment]
@pcarranza I cooked up a spec demonstrating we do indeed have a race condition: https://gist.github.com/smellsblue/19302cb81e44c0fcc4a724afe3a00a07

I've seen it fail with different exceptions, but here's a run with the exact same symptoms:
```
$ rspec spec/race_condition_spec.rb
Run options: include {:focus=>true}

All examples were filtered out; ignoring {:focus=>true}

possible race conditions
  storage
    works fine with serial code
    doesn't fail to read or write with concurrent reads and writes (FAILED - 1)

Failures:

  1) possible race conditions storage doesn't fail to read or write with concurrent reads and writes
     Failure/Error: return if version <= Gemstash::Resource::VERSION
     
     NoMethodError:
       undefined method `<=' for nil:NilClass
     # ./lib/gemstash/storage.rb:279:in `check_resource_version'
     # ./lib/gemstash/storage.rb:274:in `load_properties'
     # ./lib/gemstash/storage.rb:190:in `update_properties'
     # ./lib/gemstash/storage.rb:158:in `save'
     # ./spec/race_condition_spec.rb:11:in `block in write_thread'

Finished in 0.28243 seconds (files took 0.41599 seconds to load)
2 examples, 1 failure

Failed examples:

rspec ./spec/race_condition_spec.rb:38 # possible race conditions storage doesn't fail to read or write with concurrent reads and writes
```
:cry:

**Comment by smellsblue** [comment]
Shall I commit this spec to a new branch and we can try to tackle it...? I don't think it will be an easy one to solve though unfortunately.

**Comment by pcarranza** [comment]
How is this failing exactly @smellsblue ?

Could we do the nasty thing and just retry in case of this bug (I know... I know...)

I'm fine committing the spec, just not in master for now 😄

**Comment by smellsblue** [comment]
@pcarranza I would have to investigate closer, but I'm thinking one thread essentially has created the properties file but not fully written to it yet before another thread reads the properties. The thread doing the reading has no content in the properties file, which ends up being `false`, so our code causes it to be an empty hash: `@properties || {}` which then causes the resource version to be an empty key, causing the no method error on nil.

If this were a critical failure causing major problems, I would say a temporary hack around the problem until a better solution were in place would be fine. However, I think this is an infrequent enough case with mild enough consequences (rebundling isn't as bad as something like corrupted gems for example) that it is worth just doing the right fix from the beginning.

What is the right fix...? I'm not entirely sure. We could introduce some kind of synchronization, but that would only work if you have 1 Gemstash server pointing at 1 directory, but if it's reasonable to have 2 Gemstash servers pointing at 1 directory, then this would break down (probably not a case I would care about). Alternatively, we could use file locking. Either way, we need to be very careful not to create any deadlock scenarios.

**Comment by pcarranza** [comment]
Which is why I wonder if a simple exponential backoff and retry would just blow the problem away @smellsblue 

As you say, it's a corner case and is strange enough to find it, just retrying will give you a fully written file back (unless there is a really strange problem or really slow writing process)

**Comment by smellsblue** [comment]
@pcarranza I would worry too much about treating the symptom rather than the cause... the concurrency bug could end up biting us further down in other parts of the code or in other unexpected ways if we don't tackle the cause directly.

**Comment by indirect** [comment]
Reproduced from Slack:

> I guess what I am imagining is a lock that is almost always open, but is held when a remote request is going to be made, until the cache has been updated
> and then after that it’s fine for five threads to all serve the same cache at the same time

If something is already cached, there is no need for a lock. If something is not yet cached, other simultaneous requests should wait until the cache is filled before serving their own responses—they definitely shouldn't try to update the cache themselves at the same time, because simultaneous writes to the same resource from different threads is a recipe for disaster. 😅

**Comment by smellsblue** [comment]
This was fixed with #111
